#---
title: "ClinVar pathogenic Retina"
output: html_notebook
---

# Read in MacArthur group processed clinvar data
```{r}
library(data.table)
library(tidyverse)
library(here)
data_path <- '/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/'

#clinvar_single <- fread('gzcat ~/git/clinvar/output/b37/single/clinvar_alleles.single.b37.tsv.gz')
clinvar_single <- fread(paste('gzcat', here('processed_data/clinvar_alleles.single.b37.tsv.gz')))
```

# ID RD specific pathogenic variants
```{r}
out <- clinvar_single %>% 
  filter(pathogenic > 0 | likely_pathogenic > 0) %>% 
  filter(conflicted=='0', gold_stars > 0) %>% 
  filter(grepl('stargardt|retina|leber|usher|cone-rod|rod-cone|macula|retinitis|eye|cornea', all_traits, ignore.case=T) | 
           grepl('stargardt|retina|leber|usher|cone-rod|rod-cone|macula|retinitis|eye|cornea', xrefs, ignore.case=T)) %>% 
  mutate(ID='.', QUAL='.', FILTER='.', INFO='STATUS=PATHOGENIC_EYE', FORMAT='GT', FAUX='0/1')
dim(out)
```

# ID RD specific benign variants
```{r}
out <- rbind(out,
             clinvar_single %>% 
  filter(benign > 0) %>% 
  filter(conflicted=='0') %>% 
   filter(grepl('stargardt|retina|leber|usher|cone-rod|rod-cone|macula|retinitis|eye|cornea', all_traits, ignore.case=T) | 
           grepl('stargardt|retina|leber|usher|cone-rod|rod-cone|macula|retinitis|eye|cornea', xrefs, ignore.case=T)) %>% 
  mutate(ID='.', QUAL='.', FILTER='.', INFO='STATUS=BENIGN_EYE', FORMAT='GT', FAUX='0/1'))

out
```

# All benign
```{r}
clinvar_single %>% 
  filter(benign > 0) %>% 
  filter(conflicted=='0')

out <- rbind(out,
             clinvar_single %>% 
  filter(benign > 0, !variation_id %in% out$variation_id) %>% 
  filter(conflicted=='0') %>% 
  mutate(ID='.', QUAL='.', FILTER='.', INFO='STATUS=BENIGN', FORMAT='GT', FAUX='0/1'))
```

# Write vcf
```{r}
write('##fileformat=VCFv4.2', file=paste0(data_path,'data/clinvar_RD.vcf'))
write('##source=ClinVar', file=paste0(data_path,'data/clinvar_RD.vcf'), append = T)
write('##INFO=<ID=STATUS,Number=1,Type=String,Description=\"Non-conflicting RD pathogenic or benign in ClinVar\">', file=paste0(data_path,'data/clinvar_RD.vcf'), append = T)
write('##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">',file=paste0(data_path,'data/clinvar_RD.vcf'), append = T)
write('#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tFAUX', file=paste0(data_path,'data/clinvar_RD.vcf'), append = T)
write_tsv(out %>% mutate(pos=as.integer(pos)) %>% select(chrom, pos, ID, ref, alt, QUAL, FILTER, INFO, FORMAT, FAUX) %>% arrange(chrom, pos), path=paste0(data_path,'data/clinvar_RD.vcf'), append = T)
system(command = paste0('bgzip -f ', file=paste0(data_path,'data/clinvar_RD.vcf')))
system(command = paste0('tabix -p vcf ', file=paste0(data_path,'data/clinvar_RD.vcf.gz')))
```

# Next
Annotate vcf with `sbatch ~/git/variant_prioritization/Snakemake.wrapper.sh ~/git/eye_var_Pathogenicity/config_variant_prioritization__clinvar.yaml`

# Pull variants with gemini
`time gemini query --header -q "SELECT * from variants WHERE (aaf_esp_all < 0.01 AND aaf_1kg_all_float < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%'" clinvar_RD.PED_faux.gemini.db | bgzip > clinvar.gemini.tsv.gz`