---
title: "Model Assesement"
output: 
  html_notebook:
    theme: flatly
    toc: True
---

# Load Data
```{r}
library(tidyverse)
library(caret)
library(ModelMetrics)
library(randomForest)
library(keras)
library(tensorflow)
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/allX_2018_06_22.Rdata')
# confusion matrix maker for built models
cm_maker <- function(predictor = 'cadd_phred', data, cutoff=0.5, mode = 'prec_recall') {
  if (class(predictor)!='character'){
    print("Running in predictor is a model mode")
    new_predictions <- predict(predictor, data, type='prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Answers == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = as.factor(new_predictions$Prediction), reference = as.factor(new_predictions$Answers), mode= mode)
    out$MCC <- mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff)
  } else {
    print("Running in predictor is a precomputed column in data mode")
    new_predictions <- data 
    new_predictions$Prediction <- 'NotPathogenic'
    new_predictions[(new_predictions[,predictor] > cutoff), 'Prediction'] <- "Pathogenic"
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Status == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = as.factor(new_predictions$Prediction), reference = as.factor(new_predictions$Status), mode= mode)
    out$MCC <- mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff)
  }
  out
}


```

# Correlation of predictors
## VPaC
```{r, fig.width=6, fig.height=6}
library(superheat)

predictors<-c('ccr_pct_v1','cadd_raw','vest3_rankscore','cadd_phred','mis_z','pli','lof_z','phylop_100way','revel','hapmap2','hapmap1','n_mis','epilogos_quies','n_lof','precessive','pnull','adj_exp_lof','adj_exp_syn','dann','adj_exp_mis','syn_z','n_syn','epilogos_txwk','fitcons','m_cap_score','m_cap_rankscore','eigen_phred','eigen_raw','epilogos_tx','is_lof','eigen_pc_raw_rankscore','epilogos_reprpcwk','fathmm_mkl_coding_rankscore','metalr_score','fathmm_mkl_coding_score','metalr_rankscore','impact_severity','metasvm_rankscore','metasvm_score','epilogos_enh','genocanyon_score','fathmm_converted_rankscore','mpc','epilogos_enhg','af_exac_all','epilogos_reprpc','max_aaf_all','mutationassessor_score','gerp','polyphen_score','gerp_elements','mutationassessor_score_rankscore','stam_mean','an_exac_all','af_exac_nfe','provean_converted_rankscore','an_exac_nfe','lrt_score','lrt_omega','grantham','lrt_converted_rankscore','genocanyon_score_rankscore','an_exac_afr','an_exac_amr','an_exac_sas','epilogos_het','ac_exac_all','linsight','gno_an_popmax','exac_num_het','an_exac_eas','gno_an_all','ac_exac_nfe','mutationtaster_converted_rankscore','an_exac_oth','an_exac_fin','gno_an_nfe','gno_af_all','gno_an_afr','epilogos_tssaflnk','gno_af_popmax','epilogos_znf','segway_sum_score','aaf_esp_ea','epilogos_txflnk','provean_score','segway_mean_score','epilogos_tss','aaf_esp_all','af_exac_amr','gno_af_nfe','epilogos_enhbiv','af_exac_sas','sift_score','fathmm_score','ac_exac_amr','aaf_esp_aa','gno_ac_all','gno_af_afr','ac_exac_sas','af_exac_eas','gno_an_fin','af_exac_afr','gno_an_eas','gno_an_oth','gno_ac_nfe','gno_ac_popmax','ac_exac_eas','ac_exac_afr','epilogos_tssbiv','gno_ac_afr','vest3_score')

# remove near zero var predictors with caret's nearZeroVar
dat <- allX %>% filter(DataSet == 'Homsy') %>% select(one_of(predictors, 'VPaC_m06', 'VPaC_m15','DeepRNN'))
rem <- nearZeroVar(dat)
dat <- dat[,-rem]
superheat(cor(dat), 
          pretty.order.rows = T, 
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, 
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.5)
```

# Variable Importance
For random forest model

Mean Decrease Gini
```{r, fig.height=6, fig.width=2}
VPaC_12mtry$importance %>% data.frame() %>% rownames_to_column('Predictor') %>% 
  arrange(MeanDecreaseGini) %>% 
  mutate(Predictor = factor(Predictor, levels=unique(Predictor))) %>% 
  ggplot(aes(x=Predictor, y=MeanDecreaseGini)) +
  geom_point() +
  coord_flip() +
  theme_minimal()
```

Mean Decrease Accuracy
```{r, fig.height=6, fig.width=2}
VPaC_12mtry$importance %>% data.frame() %>% rownames_to_column('Predictor') %>% 
  arrange(MeanDecreaseAccuracy) %>% 
  mutate(Predictor = factor(Predictor, levels=unique(Predictor))) %>% 
  ggplot(aes(x=Predictor, y=MeanDecreaseAccuracy)) +
  geom_point() +
  coord_flip() +
  theme_minimal()
```

## Grimm
```{r, fig.width=6, fig.height=4}
superheat(cor(allX %>% filter(grepl('Grimm', DataSet, !pos_id %in% (allX %>% filter(DataSet == 'VPaC Train Set') %>% pull(pos_id)))) %>% select(one_of(predictors, 'VPaC_v01', 'VPaC_V02'))), 
          pretty.order.rows = T, 
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, row.dendrogram = T,
          
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.4)
```

# AUC-PR AUC-ROC Calculation Functions
```{r, fig.width=5, fig.height=1.5}
library(PRROC)
library(cowplot)

# precision recall AUC
pr_maker <- function(predictor, data, cutoff=0.5) {
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    pr.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
             scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
             curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    pr.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
             scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
             curve = T)
  }
}

# ROC AUC
roc_maker <- function(predictor, data, cutoff=0.5) {
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    roc.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
              scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
              curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    roc.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
              scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
              curve = T)
  }
}
```

# Assess Performance of Models and Existing Predictors in Test Set
```{r}
pr_data <- data.frame()
model_vec <- c('DeepRNN','DeepVPaC','VPaC_m15', 'VPaC_m06', 'ccr_pct_v1', 'revel', 'cadd_phred', 'fitcons', 'metalr_rankscore', 'fathmm_converted_rankscore')
for (i in model_vec){
  for (j in unique(allX$DataSet)){
    pr <- pr_maker(i, allX %>% filter(DataSet == j))
    out <- pr$curve[,1:2] %>% data.frame()
    colnames(out) <- c('Recall','Precision')
    out$AUC <- pr$auc.integral
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(pr$auc.integral,5),')' )
    out$DataSet <- j
    pr_data <- rbind(pr_data, out)
  }
}

roc_data <- data.frame()
for (i in model_vec){
  for (j in unique(allX$DataSet)){
    roc <- roc_maker(i, allX %>% filter(DataSet == j))
    out <- roc$curve[,1:2] %>% data.frame()
    colnames(out) <- c('FPR','Sensitivity')
    out$AUC <- roc$auc
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc,5),')' )
    out$DataSet <- j
    roc_data <- rbind(roc_data, out)
  }
}
```
## Table Data
```{r}
pr_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% 
  dplyr::select(AUC, Model, DataSet) %>% 
  #filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% 
  unique() %>% 
  mutate(AUC=round(AUC,4)) %>% 
  spread(Model, AUC) %>% 
  DT::datatable()

roc_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% dplyr::select(AUC, Model, DataSet) %>% filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% unique() %>% mutate(AUC=round(AUC,4)) %>% spread(Model, AUC) %>% DT::datatable()
```
## Dot plots of AUC
```{r, fig.height=4}
pr_data %>% dplyr::select(AUC, Model, DataSet) %>% filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% unique() %>% ggplot(aes(x=Model, y=AUC)) + geom_point(stat = 'identity') + coord_cartesian(ylim=c(0,1)) + facet_wrap(~DataSet, nrow=2) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('Precision Recall')

roc_data %>% dplyr::select(AUC, Model, DataSet) %>% filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% unique()%>% ggplot(aes(x=Model, y=AUC)) + geom_point(stat = 'identity') + coord_cartesian(ylim=c(0,1)) + facet_wrap(~DataSet, nrow=2) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('Receiver Operating Characteristic')
```
## AUC Plots
```{r, fig.height=6, fig.width=6}
pr_plot <- pr_data %>% 
  filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% 
  filter(Model != 'VPaC') %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=Model)) + 
  facet_wrap(~DataSet, ncol=1) +
  geom_step() + 
  theme_minimal() + 
  ggtitle('Precision Recall Curve') 
roc_plot <- roc_data %>%   #filter(model %in% best_models_noDC) %>% 
  filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>% 
  filter(Model != 'VPaC') %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=Model)) + 
  facet_wrap(~DataSet, ncol=1) +
  geom_step() + 
  theme_minimal() + 
  ggtitle('ROC Curve') 

# pr <- aucroc_data %>% ggplot(aes(x=Recall, y=Precision)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve')
# roc <- roc_data %>% ggplot(aes(x=FPR, y=Sensitivity)) + geom_path() + theme_minimal()  + ggtitle('ROC')

cowplot::plot_grid(roc_plot, pr_plot, ncol = 2)
```

# Find optimal cutoff for DataSet X for model Y
Plot cutoff against MCC
```{r}
mcc_view <- function(sequence = seq(0,1,by = 0.01), predictor = 'VPaC', dataset = all %>% filter(DataSet == 'Grimm ExoVar')){
  out <- data.frame(cutoff = sequence)
  mccs <- ''
  for (i in sequence){
    capture.output(mccs <- c(mccs, cm_maker(predictor, dataset, i)$MCC), file='/dev/null')
  }
  out$mcc <- mccs[2:length(mccs)]
  out$mcc[out$mcc == 'NaN'] <- 0
  out$mcc <- as.numeric(out$mcc)
  out$cutoff <- as.numeric(out$cutoff)
  out
}

suppressWarnings(mcc_view(predictor='DeepVPaC', dataset = all %>% filter(DataSet=='Grimm PredictSNP')) %>% arrange(-mcc) %>% head())
```
# OGVFB Exomes
```{r}
ogvfb$ogvfb_ML_set$VPaC <-predict(VPaC_6mtry, ogvfb$ogvfb_ML_set, type='prob')[,1]
ogvfb$ogvfb_ML_set$DiseaseClass <- factor(ogvfb$ogvfb_ML_set$DiseaseClass, levels=c('-1','Albinism','Albinism,Albinism','Developmental','RD','RD,Developmental','Stargardt,RD'))
ogvfb$ogvfb_ML_set$OVPaC <-predict(OVPaC_6mtry, ogvfb$ogvfb_ML_set, type='prob')[,1]

# OVPaC most path
ogvfb$ogvfb_ML_set %>% filter(OVPaC > 0.1) %>% 
  left_join(., ogvfb$gemini_ogvfb_candidates %>%  
              mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt), hgvsc=gsub('.*:','',hgvsc), hgvsp=gsub('.*:','',hgvsp)) %>% 
              select(pos_id, contains('hgvs'), contains('family'), gene, impact_so, num_het, num_hom_alt, filter, hgmd_overlap, clinvar_pathogenic, pubmed)) %>% 
  filter(num_het < 4, num_hom_alt < 4, filter == 'None') %>% 
  select(pos_id, gene, hgvsc, hgvsp, impact_so, hgmd_overlap, clinvar_pathogenic, pubmed, VPaC, OVPaC, ccr_pct_v1, gno_an_popmax, cadd_phred, fitcons_float, revel, family_id, family_members, family_genotypes) %>% 
  mutate(family_genotypes = gsub(',','<br/>', family_genotypes), family_members = gsub(',','<br/>', family_members)) %>% 
  arrange(-OVPaC) %>% 
  mutate_if(is.numeric, funs(round(.,2))) %>% 
  DT::datatable(escape = FALSE, rownames = FALSE)

# VPaC most path
ogvfb$ogvfb_ML_set %>% filter(VPaC > 0.1) %>% 
  left_join(., ogvfb$gemini_ogvfb_candidates %>%  
              mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt), hgvsc=gsub('.*:','',hgvsc), hgvsp=gsub('.*:','',hgvsp)) %>% 
              select(pos_id, contains('hgvs'), contains('family'), gene, impact_so, num_het, num_hom_alt, filter, hgmd_overlap, clinvar_pathogenic, pubmed)) %>% 
  filter(num_het < 4, num_hom_alt < 4, filter == 'None') %>% 
  select(pos_id, gene, hgvsc, hgvsp, impact_so, hgmd_overlap, clinvar_pathogenic, pubmed, VPaC, OVPaC, ccr_pct_v1, gno_an_popmax, cadd_phred, fitcons_float, revel, family_id, family_members, family_genotypes) %>% 
  mutate(family_genotypes = gsub(',','<br/>', family_genotypes), family_members = gsub(',','<br/>', family_members)) %>% 
  arrange(-VPaC) %>% 
  mutate_if(is.numeric, funs(round(.,2))) %>% 
  DT::datatable(escape = FALSE, rownames = FALSE)
```

# NISC 100
How do we do?
```{r}
nisc100 <- fread('gzcat /Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ddl_nisc_100_panel/ddl_nisc.gemini.tsv.gz')
left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
          nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(DeepVPaC > 0.5) %>% 
  select(Status, pos_id, hgvsc, gene, type, num_het, num_hom_alt, aaf,aaf_1kg_all_float, impact_so, DeepVPaC, clin_sig) %>% pull(Status) %>% table()

```
Not great

How about if we added a touch of info - ClinVar, Impact
```{r}
left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
          nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(DeepVPaC > 0.5 | grepl('path', clin_sig) | impact_severity == 'HIGH') %>% 
  filter(aaf < 0.15 | gno_af_popmax < 0.01) %>% 
  select(Status, pos_id, hgvsc, gene, type, num_het, num_hom_alt, aaf,aaf_1kg_all_float, impact_so, DeepVPaC, clin_sig) %>% pull(Status) %>% table()

grabbed <- 
  left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
            nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(DeepVPaC > 0.5 | grepl('path', clin_sig) | impact_severity=='HIGH') %>% 
  filter(aaf < 0.15 | gno_af_popmax < 0.01) %>% 
  select(Status, pos_id, hgvsc, gene, type, num_het, num_hom_alt, aaf,aaf_1kg_all_float, impact_so, DeepVPaC, clin_sig) %>% pull(pos_id)



left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
          nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(Status == 'Pathogenic', !pos_id %in% grabbed) %>% 
  select(Status, pos_id, hgvsc, gene, type, num_het, num_hom_alt, impact_so, DeepVPaC, clin_sig) %>% 
  DT::datatable()

```


High ranking Deep VPaC that are Not Pathogenic in NISC 100
```{r}
left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
          nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(DeepVPaC > 0.9) %>% 
  filter(Status == 'NotPathogenic') %>% 
  select(Status, pos_id, hgvsc, gene, type, num_het, num_hom_alt, aaf,aaf_1kg_all_float, impact_so, DeepVPaC, clin_sig) %>% 
  DT::datatable()

left_join(all %>% filter(DataSet=='DDL NISC RD Cohort') %>% select(pos_id, Status, DeepVPaC),
          nisc100 %>% mutate(pos_id = paste0(chrom,':',end, '_', ref, '_', alt))) %>% 
  filter(DeepVPaC > 0.9) %>% 
  filter(Status == 'NotPathogenic') %>% 
  select(hgvsc, gene, type, aaf_1kg_all_float, impact_so, DeepVPaC, clin_sig) %>% 
  DT::datatable()

```
}}
