---
title: "Model Assesement"
output: html_notebook
---
# Load processed data and models
From scripts/build_models.R run on biowulf2
```{r}
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/model_data.Rdata')
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/VPaC_6mtry_RF.Rdata')
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/OVPaC_RF.Rdata')
# ogvfb cohort
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/ogvfb_exome_cohort.Rdata')
# grimm
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/grimm_ML.Rdata')
# unifun
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/unifun_ML.Rdata')
```


# ConfusionMatrix your model(s)
```{r}
library(tidyverse)
library(caret)
library(ModelMetrics)
library(randomForest)
# confusion matrix maker for built models
cm_maker <- function(predictor = 'cadd_phred', data, cutoff=0.5, mode = 'prec_recall') {
  if (class(predictor)!='character'){
    print("Running in predictor is a model mode")
    new_predictions <- predict(predictor, data, type='prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Answers == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = new_predictions$Prediction, reference = new_predictions$Answers, mode= mode)
    print(mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff))
  } else {
    print("Running in predictor is a precomputed column in data mode")
    new_predictions <- data 
    new_predictions$Prediction <- 'NotPathogenic'
    new_predictions[(new_predictions[,predictor] > cutoff), 'Prediction'] <- "Pathogenic"
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Status == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = new_predictions$Prediction, reference = new_predictions$Status, mode= mode)
    print(mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff))
  }
  out
}


cm_maker('revel', grimm$grimm_ML_set, cutoff=0.5)
cm_maker(VPaC, grimm$grimm_ML_set, cutoff=0.01)
```

# Merge all datasets and calculated OVPaC and VPaC scores
```{r}
useful_cols <- c('pos_id', 'Status', 'DiseaseClass', 'is_lof','impact_severity','mis_z','ccr_pct_v1','cadd_phred','phylop_100way','n_mis','revel','fitcons_float','precessive','n_lof','m_cap_rankscore','dann','vest3_rankscore','n_syn','pnull','pli','lof_z','fathmm_mkl_coding_rankscore','an_exac_all','eigen_pc_raw_rankscore','gerp_elements','mutationassessor_score_rankscore','mpc','metasvm_rankscore','polyphen_score','metalr_rankscore','lrt_converted_rankscore','genocanyon_score_rankscore','mutationtaster_converted_rankscore','gno_an_popmax','grantham','max_aaf_all','ac_exac_all','fathmm_converted_rankscore','aaf_esp_all','linsight', 'sift_score','ac_exac_sas')

all <- rbind(model_data$ML_set__general_TT$train_set %>% dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'VPaC Train Set'), 
             model_data$ML_set__general_TT$test_set %>% dplyr::select(one_of(useful_cols)) %>%mutate(DataSet = 'VPaC Test Set'),
             model_data$ML_set__eye_TT$train_set %>% dplyr::select(one_of(useful_cols)) %>%mutate(DataSet = 'OVPaC Train Set'),
             model_data$ML_set__eye_TT$test_set %>% dplyr::select(one_of(useful_cols)) %>%mutate(DataSet = 'OVPaC Test Set'),
             model_data$ML_set__other_TT$train_set %>% dplyr::select(one_of(useful_cols)) %>%mutate(DataSet = 'Remainders'),
             model_data$ML_set__other_TT$test %>% dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'ClinVar nonHC Pathogenic and remaining gnomAD'),
             grimm$grimm_ML_set %>% 
               inner_join(., grimm$grimm_orig %>% mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt)) %>% 
                            dplyr::select(pos_id, source) %>% 
                            filter(grepl('humvar', source))) %>% 
               dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'Grimm HumVar'),
             grimm$grimm_ML_set %>% 
               inner_join(., grimm$grimm_orig %>% mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt)) %>% 
                            dplyr::select(pos_id, source) %>% 
                            filter(grepl('exovar', source))) %>% 
               dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'Grimm ExoVar'),
             grimm$grimm_ML_set %>% 
               inner_join(., grimm$grimm_orig %>% mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt)) %>% 
                            dplyr::select(pos_id, source) %>% 
                            filter(grepl('predictS', source))) %>% 
               dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'Grimm PredictSNP'),
             grimm$grimm_ML_set %>% 
               inner_join(., grimm$grimm_orig %>% mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt)) %>% 
                            dplyr::select(pos_id, source) %>% 
                            filter(grepl('swissv', source))) %>% 
               dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'Grimm SwissVar'),
             grimm$grimm_ML_set %>% 
               inner_join(., grimm$grimm_orig %>% mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt)) %>% 
                            dplyr::select(pos_id, source) %>% 
                            filter(grepl('varibench', source))) %>% 
               dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'Grimm VariBench'),
             unifun$unifun_ML_set %>% dplyr::select(one_of(useful_cols)) %>% mutate(DataSet = 'UniFun'))
all$OVPaC <-predict(OVPaC, all, type='prob')[,1]
all$VPaC <-predict(VPaC, all, type='prob')[,1]
```

# Correlation of predictors
```{r, fig.width=6, fig.height=5}
library(superheat)
library(randomForest)

predictors <- c('is_lof','impact_severity','mis_z','ccr_pct_v1','cadd_phred','phylop_100way','n_mis','revel','fitcons_float','precessive','n_lof','m_cap_rankscore','dann','vest3_rankscore','n_syn','pnull','pli','lof_z','fathmm_mkl_coding_rankscore','an_exac_all','eigen_pc_raw_rankscore','gerp_elements','mutationassessor_score_rankscore','mpc','metasvm_rankscore','polyphen_score','metalr_rankscore','lrt_converted_rankscore','genocanyon_score_rankscore','mutationtaster_converted_rankscore','gno_an_popmax','grantham','max_aaf_all','ac_exac_all','fathmm_converted_rankscore','aaf_esp_all','linsight', 'sift_score','ac_exac_sas')

superheat(cor(all %>% filter(DataSet=='OVPaC Test Set') %>% select(one_of(predictors, 'OVPaC', 'VPaC'))), 
          pretty.order.rows = T, 
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, 
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.5)
```

# AUC-PR AUC-ROC Calculation Functions
```{r, fig.width=5, fig.height=1.5}
library(PRROC)
library(cowplot)

# precision recall AUC
pr_maker <- function(predictor, data, cutoff=0.5) {
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    pr.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
             scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
             curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    pr.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
             scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
             curve = T)
  }
}

# ROC AUC
roc_maker <- function(predictor, data, cutoff=0.5) {
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    roc.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
              scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
              curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    roc.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
              scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
              curve = T)
  }
}
```

# Assess Performance of Models and Existing Predictors in Test Set
```{r}
pr_data <- data.frame()
for (i in c('OVPaC', 'VPaC', 'revel', 'cadd_phred', 'fitcons_float', 'metalr_rankscore', 'fathmm_converted_rankscore')){
  for (j in unique(all$DataSet)){
    pr <- pr_maker(i, all %>% filter(DataSet == j))
    out <- pr$curve[,1:2] %>% data.frame()
    colnames(out) <- c('Recall','Precision')
    out$AUC <- pr$auc.integral
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(pr$auc.integral,2),')' )
    out$DataSet <- j
    pr_data <- rbind(pr_data, out)
  }
}

roc_data <- data.frame()
for (i in c('OVPaC', 'VPaC', 'revel', 'cadd_phred', 'fitcons_float', 'metalr_rankscore', 'fathmm_converted_rankscore')){
  for (j in unique(all$DataSet)){
    roc <- roc_maker(i, all %>% filter(DataSet == j))
    out <- roc$curve[,1:2] %>% data.frame()
    colnames(out) <- c('FPR','Sensitivity')
    out$AUC <- roc$auc
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc,2),')' )
    out$DataSet <- j
    roc_data <- rbind(roc_data, out)
  }
}
```
## Table Data
```{r}
pr_data %>% dplyr::select(AUC, Model, DataSet) %>% filter(grepl('Grimm|ClinVar|Test|UniFun', DataSet)) %>% unique() %>% mutate(AUC=round(AUC,2)) %>% spread(Model, AUC)

roc_data %>% dplyr::select(AUC, Model, DataSet) %>% filter(grepl('Grimm|ClinVar|Test|UniFun', DataSet)) %>% unique() %>% mutate(AUC=round(AUC,2)) %>% spread(Model, AUC)
```
## Dot plots of AUC
```{r}
pr_data %>% dplyr::select(AUC, Model, DataSet) %>% unique() %>% ggplot(aes(x=Model, y=AUC)) + geom_bar(stat = 'identity') + coord_cartesian(ylim=c(0,1)) + facet_wrap(~DataSet) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('AUC PR')

roc_data %>% dplyr::select(AUC, Model, DataSet) %>% unique() %>% ggplot(aes(x=Model, y=AUC)) + geom_bar(stat = 'identity') + coord_cartesian(ylim=c(0,1)) + facet_wrap(~DataSet) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('AUC ROC')
```
## AUC Plots
```{r}
pr_plot <- pr_data %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=Model)) + 
  facet_wrap(~DataSet, ncol=1) +
  geom_step() + 
  theme_minimal() + 
  ggtitle('Precision Recall Curve') 
roc_plot <- roc_data %>%   #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=Model)) + 
  facet_wrap(~DataSet, ncol=1) +
  geom_step() + 
  theme_minimal() + 
  ggtitle('ROC Curve') 

# pr <- aucroc_data %>% ggplot(aes(x=Recall, y=Precision)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve')
# roc <- roc_data %>% ggplot(aes(x=FPR, y=Sensitivity)) + geom_path() + theme_minimal()  + ggtitle('ROC')

cowplot::plot_grid(roc_plot, pr_plot, ncol = 2)
```

Non ocular pathogenic
```{r}
model_names <- c('rfFit_noDC', 'bglmFit_noDC', 'caddFit', 'dannFit', 'revelFit')
aucroc_data <- data.frame()
for (i in model_names){
  print(model_run[[i]]$method)
  aucroc <- aucroc_maker(model_run[[i]], model_run$ML_set_dummy__secondary)
  out <- aucroc$curve[,1:2] %>% data.frame()
  colnames(out) <- c('Recall','Precision')
  out$AUC <- aucroc$auc.integral
  out$model <- i
  out$'Model (AUC)' <- paste0(i, ' (',round(aucroc$auc.integral,2),')' )
  aucroc_data <- rbind(aucroc_data, out)
}

# roc_data <- data.frame()
# for (i in model_names){
#   print(model_run_core[[i]]$method)
#   roc <- roc_maker(model_run_core[[i]], model_run_core$ML_set_dummy__secondary)
#   out <- roc$curve[,1:2] %>% data.frame()
#   colnames(out) <- c('FPR','Sensitivity')
#   out$model <- i
#   out$AUC <- roc$auc
#   out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc, 2),')' )
#   roc_data <- rbind(roc_data, out)
# }

pr <- aucroc_data %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve') + ggsci::scale_colour_aaas()
# roc <- roc_data %>% 
#   # filter(model %in% best_models_noDC) %>% 
#   ggplot(aes(x=FPR, y=Sensitivity, colour=`Model (AUC)`)) + geom_path() + theme_minimal()  + ggtitle('ROC')+ ggsci::scale_colour_aaas()

# pr <- aucroc_data %>% ggplot(aes(x=Recall, y=Precision)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve')
# roc <- roc_data %>% ggplot(aes(x=FPR, y=Sensitivity)) + geom_path() + theme_minimal()  + ggtitle('ROC')

pr
```


# Dive into fails
First load metadata for UK10K and full dataset
```{r}
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/uk10k_gemini_rare_variants.Rdata')
metadata <- readxl::read_excel(path='~/git/EGA_EGAD00001002656_NGS_reanalyze/data/1-s2.0-S0002929716305274-mmc3.xlsx') %>% mutate(Sample=Patient)

predictor <- function(model, data_dummy, source){
  # returns variant id to match to full data frame and pathogenic score
  # Set source to only return predictions from a certain source
  # variant id is only unique to each source, as they are separate gemini db
  # add empty -1 columns to data_dummy if predictor not present
  columns_needed <- model$trainingData %>% colnames()
  columns_needed <- columns_needed[2:length(columns_needed)]
  missing_columns <- columns_needed[!columns_needed %in% colnames(data_dummy)]
  data_dummy[,missing_columns] <- -1
  # predict
  predictions <- predict(model, data_dummy, type = 'prob')
  data_dummy <- cbind(predictions, data_dummy)
  return(data_dummy %>% filter(Source == source) %>% select(variant_id, Pathogenic))
}
```

## OVPaC labels as pathogenic, but UK10K not labeled path
Validation subset
```{r}

left_join(predictor(model_run$rfFit, model_run$validate_set, 'UK10K'), uk10k_gemini_rare_variants) %>% 
  filter(filter=='None') %>% 
  separate(hgvsc, c('tx_c', 'hgvsc'), ':') %>% 
  separate(hgvsp, c('tx_p', 'hgvsp'), ':') %>% 
  filter(Pathogenic > 0.5, Status=='NotPathogenic') %>% 
  select(OVPaC=Pathogenic, Status, impact, gene, hgvsc, hgvsp, variant_samples, het_samples, hom_alt_samples, cadd_phred, revel, gno_ac_all, ac_exac_all, exac_num_hom_alt) %>% 
  rowwise() %>% 
  mutate(variant_samples = str_split(variant_samples, pattern=',')) %>% 
  unnest(variant_samples) %>% 
  mutate(Sample = variant_samples) %>% 
  select(-variant_samples) %>% 
  left_join(., metadata %>% select(Sample, Phenotype, Gene, Variant_HGVSp)) %>% 
  DT::datatable()
```

## OVPaC labels as non-pathogenic, but UK10K labeled path
Validation subset
```{r}

left_join(predictor(model_run$rfFit, model_run$validate_set, 'UK10K'), uk10k_gemini_rare_variants) %>% 
  filter(filter=='None') %>% 
  separate(hgvsc, c('tx_c', 'hgvsc'), ':') %>% 
  separate(hgvsp, c('tx_p', 'hgvsp'), ':') %>% 
  filter(Pathogenic < 0.5, Status=='Pathogenic') %>% 
  select(OVPaC=Pathogenic, impact, gene, hgvsc, hgvsp, het_samples, hom_alt_samples, cadd_phred, revel, linsight, ac_exac_all, exac_num_hom_alt, hgmd_overlap, clin_sig) %>% 
  unique() %>% 
  DT::datatable()

```
