---
title: "Model Assesement"
output: html_notebook
---
# Load processed data and models
From scripts/build_models.R run on biowulf2
```{r}
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/model_data.Rdata')
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/VPaC_RF.Rdata')
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/OVPaC_RF.Rdata')
# ogvfb cohort
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/ogvfb_exome_cohort.Rdata')
# unifun
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/unifun_ML.Rdata')
```


# ConfusionMatrix your model(s)
```{r}
library(tidyverse)
library(caret)
cm_maker <- function(model, data, cutoff=0.5, mode = 'prec_recall') {
  new_predictions <- predict(model, data, type='prob') %>% data.frame() %>% 
    mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
  confusionMatrix(data = new_predictions$Prediction, reference = new_predictions$Answers, mode= mode)
}

cm_maker(VPaC, model_data$ML_set__general_dummy_TT$test_set)
#cm_maker(VPaC, model_data$ML_set__general_dummy_TT$train_set)

cm_maker(OVPaC, model_data$ML_set__eye_TT$test_set)
#cm_maker(OVPaC, model_data$ML_set__eye_TT$train_set)
for (i in model_names){
  print(i)
  print(cm_maker(model_run[[i]], model_run$validate_set))
}
```

# Best predictors (correlation)
```{r, fig.width=6, fig.height=5}
library(superheat)
library(randomForest)
all <- rbind(model_data$ML_set__general_TT$train_set, model_data$ML_set__general_TT$test_set)
all$OVPaC <-predict(OVPaC, all, type='prob')[,1]
all$VPaC <-predict(VPaC, all, type='prob')[,1]

superheat(cor(all %>% select_(.dots=c(model_run$most_imp_predictors, 'OVPaC', 'VPaC'))), 
          pretty.order.rows = T, 
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, 
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.5)
```

# AUC ROC
```{r, fig.width=5, fig.height=1.5}
library(PRROC)
library(cowplot)

# precision recall AUC
aucroc_maker <- function(model, data, cutoff=0.5) {
  new_predictions <- predict(model, data, type = 'prob') %>% data.frame() %>% 
    mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
  pr.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
           scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
           curve = T)
}

# ROC AUC
roc_maker <- function(model, data, cutoff=0.5) {
  new_predictions <- predict(model, data, type = 'prob') %>%  data.frame() %>% 
    mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
  roc.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
            scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
            curve = T)
}

# best_models_noDC <- c('rfFit_noDC','avNNetFit_noDC','bglmFit_noDC')
# best_models <- c('rfFit','avNNetFit','bglmFit')

model_names <- c('OVPaC','VPaC')#best_models_noDC

aucroc_data <- data.frame()
for (i in model_names){
  #print(model_run[[i]]$method)
  aucroc <- aucroc_maker(model_run[[i]], model_data$ML_set__other_TT$test_set)
  out <- aucroc$curve[,1:2] %>% data.frame()
  colnames(out) <- c('Recall','Precision')
  out$AUC <- aucroc$auc.integral
  out$model <- i
  out$'Model (AUC)' <- paste0(i, ' (',round(aucroc$auc.integral,2),')' )
  aucroc_data <- rbind(aucroc_data, out)
}

roc_data <- data.frame()
for (i in model_names){
  #print(model_run[[i]]$method)
  roc <- roc_maker(model_run[[i]], model_data$ML_set__eye_TT$test_set)
  out <- roc$curve[,1:2] %>% data.frame()
  colnames(out) <- c('FPR','Sensitivity')
  out$model <- i
  out$AUC <- roc$auc
  out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc, 2),')' )
  roc_data <- rbind(roc_data, out)
}

pr <- aucroc_data %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve') + ggsci::scale_colour_aaas()
roc <- roc_data %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=`Model (AUC)`)) + geom_path() + theme_minimal()  + ggtitle('ROC')+ ggsci::scale_colour_aaas()

# pr <- aucroc_data %>% ggplot(aes(x=Recall, y=Precision)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve')
# roc <- roc_data %>% ggplot(aes(x=FPR, y=Sensitivity)) + geom_path() + theme_minimal()  + ggtitle('ROC')

cowplot::plot_grid(roc, pr)
```

Non ocular pathogenic
```{r}
model_names <- c('rfFit_noDC', 'bglmFit_noDC', 'caddFit', 'dannFit', 'revelFit')
aucroc_data <- data.frame()
for (i in model_names){
  print(model_run[[i]]$method)
  aucroc <- aucroc_maker(model_run[[i]], model_run$ML_set_dummy__secondary)
  out <- aucroc$curve[,1:2] %>% data.frame()
  colnames(out) <- c('Recall','Precision')
  out$AUC <- aucroc$auc.integral
  out$model <- i
  out$'Model (AUC)' <- paste0(i, ' (',round(aucroc$auc.integral,2),')' )
  aucroc_data <- rbind(aucroc_data, out)
}

# roc_data <- data.frame()
# for (i in model_names){
#   print(model_run_core[[i]]$method)
#   roc <- roc_maker(model_run_core[[i]], model_run_core$ML_set_dummy__secondary)
#   out <- roc$curve[,1:2] %>% data.frame()
#   colnames(out) <- c('FPR','Sensitivity')
#   out$model <- i
#   out$AUC <- roc$auc
#   out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc, 2),')' )
#   roc_data <- rbind(roc_data, out)
# }

pr <- aucroc_data %>% 
  #filter(model %in% best_models_noDC) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve') + ggsci::scale_colour_aaas()
# roc <- roc_data %>% 
#   # filter(model %in% best_models_noDC) %>% 
#   ggplot(aes(x=FPR, y=Sensitivity, colour=`Model (AUC)`)) + geom_path() + theme_minimal()  + ggtitle('ROC')+ ggsci::scale_colour_aaas()

# pr <- aucroc_data %>% ggplot(aes(x=Recall, y=Precision)) + geom_path() + theme_minimal() + ggtitle('Precision Recall Curve')
# roc <- roc_data %>% ggplot(aes(x=FPR, y=Sensitivity)) + geom_path() + theme_minimal()  + ggtitle('ROC')

pr
```


# Dive into fails
First load metadata for UK10K and full dataset
```{r}
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/uk10k_gemini_rare_variants.Rdata')
metadata <- readxl::read_excel(path='~/git/EGA_EGAD00001002656_NGS_reanalyze/data/1-s2.0-S0002929716305274-mmc3.xlsx') %>% mutate(Sample=Patient)

predictor <- function(model, data_dummy, source){
  # returns variant id to match to full data frame and pathogenic score
  # Set source to only return predictions from a certain source
  # variant id is only unique to each source, as they are separate gemini db
  # add empty -1 columns to data_dummy if predictor not present
  columns_needed <- model$trainingData %>% colnames()
  columns_needed <- columns_needed[2:length(columns_needed)]
  missing_columns <- columns_needed[!columns_needed %in% colnames(data_dummy)]
  data_dummy[,missing_columns] <- -1
  # predict
  predictions <- predict(model, data_dummy, type = 'prob')
  data_dummy <- cbind(predictions, data_dummy)
  return(data_dummy %>% filter(Source == source) %>% select(variant_id, Pathogenic))
}
```

## OVPaC labels as pathogenic, but UK10K not labeled path
Validation subset
```{r}

left_join(predictor(model_run$rfFit, model_run$validate_set, 'UK10K'), uk10k_gemini_rare_variants) %>% 
  filter(filter=='None') %>% 
  separate(hgvsc, c('tx_c', 'hgvsc'), ':') %>% 
  separate(hgvsp, c('tx_p', 'hgvsp'), ':') %>% 
  filter(Pathogenic > 0.5, Status=='NotPathogenic') %>% 
  select(OVPaC=Pathogenic, Status, impact, gene, hgvsc, hgvsp, variant_samples, het_samples, hom_alt_samples, cadd_phred, revel, gno_ac_all, ac_exac_all, exac_num_hom_alt) %>% 
  rowwise() %>% 
  mutate(variant_samples = str_split(variant_samples, pattern=',')) %>% 
  unnest(variant_samples) %>% 
  mutate(Sample = variant_samples) %>% 
  select(-variant_samples) %>% 
  left_join(., metadata %>% select(Sample, Phenotype, Gene, Variant_HGVSp)) %>% 
  DT::datatable()
```

## OVPaC labels as non-pathogenic, but UK10K labeled path
Validation subset
```{r}

left_join(predictor(model_run$rfFit, model_run$validate_set, 'UK10K'), uk10k_gemini_rare_variants) %>% 
  filter(filter=='None') %>% 
  separate(hgvsc, c('tx_c', 'hgvsc'), ':') %>% 
  separate(hgvsp, c('tx_p', 'hgvsp'), ':') %>% 
  filter(Pathogenic < 0.5, Status=='Pathogenic') %>% 
  select(OVPaC=Pathogenic, impact, gene, hgvsc, hgvsp, het_samples, hom_alt_samples, cadd_phred, revel, linsight, ac_exac_all, exac_num_hom_alt, hgmd_overlap, clin_sig) %>% 
  unique() %>% 
  DT::datatable()
 
```
