---
title: "Build and process OGVFB gemini genotypes"
output: html_notebook
---

# Load in metadata

```{r}
library(tidyverse)
metadata <- read_delim('~/git/NGS_db/master.ped',delim=' ', col_names = F) 
colnames(metadata) <- c('Family','Sample','Paternal','Maternal','Gender','Phenotype')
metadata
```

# ID families with proband
```{r}
fams <- metadata %>% filter(Phenotype==2) %>% pull(Family) %>% unique()
fams
```

# Create gemini query for each family
Phenotype==2 are the probands, Phenotype==1 are the unaffected
```{r}
library(here)
#hom 

query_writer <- function(family_name, gemini_db = '2018-03-23_OGVFB_cohort.gemini.db') {
  de_novo_query <- paste0("gemini de_novo --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all_float < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.denovo')
  
  ar_query <- paste0("gemini autosomal_recessive --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all_float < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.ar')
  
  comphet_query <- paste0("gemini comp_hets --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all_float < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.ch')
  
  queries <- c(de_novo_query, ar_query, comphet_query)
  return(queries)
}
#cat(het_query)

swarm_calls <- ''
for (i in fams){swarm_calls <- c(swarm_calls, query_writer(i))}


write(swarm_calls, file = here('scripts/gemini_query_calls_OGVFB.swarm'))
```

# Run above with swarm in /data/mcgaugheyd/projects/nei/mcgaughey/eye_var_Pathogenicity/data
```{bash, eval=F}
swarm -f ~/git/eye_var_Pathogenicity/scripts/gemini_query_calls_OGVFB.swarm --module=gemini
```


# Load in result from gemini query above
```{r, results='hide'}
library(data.table)
gemini_call_files <- list.files('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb_cohort_calls/')
gemini_call_df <- gemini_call_files %>% data.frame() %>% separate('.', c('family','gemini','type'), sep = '\\.', remove=F)

gemini_ogvfb_candidates <- data.frame()
for (i in seq(1:nrow(gemini_call_df))){
  print(gemini_call_df[i,])
  input <- read_tsv(paste0('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb_cohort_calls/', gemini_call_df[i,1]))
  if (nrow(input) > 0 & gemini_call_df[i,4] != 'ch'){
    input$Gemini_Test <- gemini_call_df[i,4]
    input$Family <- gemini_call_df[i,2]
    gemini_ogvfb_candidates <- rbind(gemini_ogvfb_candidates, input)
  }
}

gemini_ogvfb_candidates_CH <- data.frame()
for (i in seq(1:nrow(gemini_call_df))){
  print(gemini_call_df[i,])
  input <- read_tsv(paste0('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb_cohort_calls/', gemini_call_df[i,1]))
  if (nrow(input) > 0 & gemini_call_df[i,4] == 'ch'){
    input$Gemini_Test <- gemini_call_df[i,4]
    input$Family <- gemini_call_df[i,2]
    gemini_ogvfb_candidates_CH <- rbind(gemini_ogvfb_candidates_CH, input)
  }
}
```

# Reshape
We have a column (hom_alt or het)_samples which has a comma separate list of samples with the genotype. I think we should split this up and get a new row for each sample. 
```{r}
# hom_long <- hom %>% 
#   mutate(Sample = str_split(hom_alt_samples, ","), Genotype = 'Hom') %>%  # split with str_split to create a list
#   unnest(Sample, .drop=F) # unnest will spread the list, creating a row for each entry
# 
# het_long <- het %>% 
#   mutate(Sample = str_split(het_samples, ","), Genotype = 'Het') %>%  # split with str_split to create a list
#   unnest(Sample, .drop=F) # unnest will spread the list, creating a row for each entry
# 
# 
# all <- rbind(hom_long, het_long)
```


# Label samples as having RD (eye disease) or not
```{r}
# all <- left_join(all, metadata %>% select(Sample, Phenotype), by='Sample')
```


# Prep for model assessment
```{r}
library(dummies)
library(caret)
all_processed <- gemini_ogvfb_candidates %>% 
  separate(gene_eyediseaseclass, c('RDGene','DiseaseClass'), sep='_') %>%  #split off RD disease type
  select(-RDGene) %>% 
  mutate(impact_severity = case_when(impact_severity == 'HIGH' ~ 3, # convert to integer 
                                     impact_severity == 'MED' ~ 2, 
                                     TRUE ~ 1),
         genesplicer = case_when(genesplicer == "" ~ 'No',
                                 grepl('^gain', genesplicer) ~ 'Gain',
                                 grepl('^loss', genesplicer) ~ 'Loss',
                                 grepl('^diff', genesplicer) ~ 'Diff',
                                 TRUE ~ 'Else')) %>%  
  mutate_at(vars(matches('ac_|an_|^n_')), funs(as.integer(.))) %>% # convert columns with ac_|whatever to integer (ac is allele count)
  mutate_at(vars(matches('af_|dann|revel|mpc|gerp|polyphen_score|sift_score|fitcons_float|gerp_elements|^adj|_z$|^pli$|^pnull$|precessive|^phylop|grantham|fathmm|linsight|metalr_score|lrt_converted_rankscore|metasvm|eigen|mutationtaster|provean')), funs(as.numeric(.))) %>%  # af is allele frequency
  select(variant_id, is_exonic, is_coding, is_lof, is_splicing, impact_severity, polyphen_score, sift_score, dann, gerp_elements, DiseaseClass, mpc, revel, max_aaf_all, gno_ac_afr, gno_ac_eas, gno_ac_all, gno_ac_popmax, ac_exac_sas, ac_exac_fin, aaf_1kg_all_float, aaf_esp_all, ac_exac_all, ac_exac_amr, ac_exac_oth, gno_af_all, gno_an_popmax, an_exac_all, af_exac_all, fitcons_float, lof_z:precessive, phylop_100way, grantham, cadd_phred, fathmm_mkl_coding_score, linsight, metalr_score, lrt_converted_rankscore, metasvm_rankscore, eigen_phred, mutationtaster_score, provean_converted_rankscore, genesplicer, spliceregion) %>% 
  filter(max_aaf_all < 0.01) %>% 
  unique() # remove any common variants

# fill missing with -1
all_processed[is.na(all_processed)] <- -1

temp <- all_processed %>% dplyr::select(-variant_id)
temp <- dummy.data.frame(temp %>% data.frame(), sep='_') 
ogvfb_ML_set_dummy <- temp %>% mutate(variant_id = all_processed$variant_id)
# center scale (don't bother, set during training)
#ML_set_dummy_CS <- preProcess(ML_set_dummy, method = c('center','scale')) %>% predict(., ML_set_dummy)

```
Save for further analysis
```{r}
ogvfb <- list()
ogvfb$all_processed  <- all_processed
ogvfb$ogvfb_ML_set_dummy <- ogvfb_ML_set_dummy

output_file <- '/Volumes/Arges_NFS/PROJECTS/mcgaughey/EGA_EGAD00001002656_NGS_reanalyze/clean_data/ogvfb_exome_cohort.Rdata'
if(!file.exists(output_file)){
  save(ogvfb, file = output_file)
}
```

# Label variants with model!
```{r}
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/model_run_2018_03_26.Rdata')

predictor <- function(model, data_dummy, source){
  # returns variant id to match to full data frame and pathogenic score
  # Set source to only return predictions from a certain source
    # variant id is only unique to each source, as they are separate gemini db
  # add empty -1 columns to data_dummy if predictor not present
  columns_needed <- model$trainingData %>% colnames()
  columns_needed <- columns_needed[2:length(columns_needed)]
  missing_columns <- columns_needed[!columns_needed %in% colnames(data_dummy)]
  data_dummy[,missing_columns] <- -1
  # predict
  predictions <- predict(model, data_dummy, type = 'prob')
  data_dummy <- cbind(predictions, data_dummy)
  return(data_dummy %>% filter(Source == source) %>% select(variant_id, Pathogenic))
}

# rfFit
left_join(predictor(model_run$rfFit, ogvfb_ML_set_dummy) %>% filter(Pathogenic>0.4), gemini_ogvfb_candidates, by='variant_id') %>% 
  select(variant_id:spliceregion, family_id:Family) %>% 
  filter(filter=='None') %>% 
  DT::datatable()

# rfFit without disease class
left_join(predictor(model_run$rfFit_noDC, ogvfb_ML_set_dummy) %>% filter(Pathogenic>0.4), gemini_ogvfb_candidates, by='variant_id') %>% 
  select(variant_id:spliceregion, family_id:Family) %>% 
  DT::datatable()
```

