`---
title: "Build and process OGVFB gemini genotypes"
output: html_notebook
---

# Load in metadata

```{r}
library(tidyverse)
metadata <- read_delim('~/git/NGS_db/master.ped',delim=' ', col_names = F) 
colnames(metadata) <- c('Family','Sample','Paternal','Maternal','Gender','Phenotype')
metadata
```

# ID families with proband
```{r}
fams <- metadata %>% filter(Phenotype==2) %>% pull(Family) %>% unique()
fams
```

# Create gemini query for each family
Phenotype==2 are the probands, Phenotype==1 are the unaffected
```{r}
library(here)
#hom 

query_writer <- function(family_name, gemini_db = 'VCFs.GATK.PED_master.gemini.db') {
  de_novo_query <- paste0("gemini de_novo --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.denovo')
  
  ar_query <- paste0("gemini autosomal_recessive --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.ar')
  
  comphet_query <- paste0("gemini comp_hets --filter \"aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%' \" --families ", family_name, ' ', gemini_db, ' > ogvfb_cohort_calls/', family_name, '.gemini.ch')
  
  queries <- c(de_novo_query, ar_query, comphet_query)
  return(queries)
}
#cat(het_query)

swarm_calls <- ''
for (i in fams){swarm_calls <- c(swarm_calls, query_writer(i))}


write(swarm_calls, file = here('scripts/gemini_query_calls_OGVFB.swarm'))
```

# Run above with swarm in /data/mcgaugheyd/projects/nei/mcgaughey/eye_var_Pathogenicity/data
```{bash, eval=F}
swarm -f ~/git/eye_var_Pathogenicity/scripts/gemini_query_calls_OGVFB.swarm --module=gemini
```


# Load in result from gemini query above
```{r, results='hide'}
library(data.table)
gemini_call_files <- list.files('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb/ogvfb_cohort_calls/')
gemini_call_df <- gemini_call_files %>% data.frame() %>% separate('.', c('family','gemini','type'), sep = '\\.', remove=F)

gemini_ogvfb_candidates <- data.frame()
for (i in seq(1:nrow(gemini_call_df))){
  #print(gemini_call_df[i,])
  input <- read_tsv(paste0('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb/ogvfb_cohort_calls/', gemini_call_df[i,1]))
  if (nrow(input) > 0 & gemini_call_df[i,4] != 'ch'){
    input$Gemini_Test <- gemini_call_df[i,4]
    input$Family <- gemini_call_df[i,2]
    gemini_ogvfb_candidates <- rbind(gemini_ogvfb_candidates, input)
  }
}

gemini_ogvfb_candidates_CH <- data.frame()
for (i in seq(1:nrow(gemini_call_df))){
  #print(gemini_call_df[i,])
  input <- read_tsv(paste0('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/ogvfb/ogvfb_cohort_calls/', gemini_call_df[i,1]))
  if (nrow(input) > 0 & gemini_call_df[i,4] == 'ch'){
    input$Gemini_Test <- gemini_call_df[i,4]
    input$Family <- gemini_call_df[i,2]
    gemini_ogvfb_candidates_CH <- rbind(gemini_ogvfb_candidates_CH, input)
  }
}
```

# Reshape
We have a column (hom_alt or het)_samples which has a comma separate list of samples with the genotype. I think we should split this up and get a new row for each sample. 
```{r}
# hom_long <- hom %>% 
#   mutate(Sample = str_split(hom_alt_samples, ","), Genotype = 'Hom') %>%  # split with str_split to create a list
#   unnest(Sample, .drop=F) # unnest will spread the list, creating a row for each entry
# 
# het_long <- het %>% 
#   mutate(Sample = str_split(het_samples, ","), Genotype = 'Het') %>%  # split with str_split to create a list
#   unnest(Sample, .drop=F) # unnest will spread the list, creating a row for each entry
# 
# 
# all <- rbind(hom_long, het_long)
```


# Label samples as having RD (eye disease) or not
```{r}
# all <- left_join(all, metadata %>% select(Sample, Phenotype), by='Sample')
```


# Prep for model assessment
```{r}
numeric_predictors <- unique(c('ccr_pct_v1','cadd_raw','vest3_rankscore','cadd_phred','mis_z','pli','lof_z','phylop_100way','revel','hapmap2','hapmap1','n_mis','epilogos_quies','n_lof','precessive','pnull','adj_exp_lof','adj_exp_syn','dann','adj_exp_mis','syn_z','n_syn','epilogos_txwk','fitcons','m_cap_score','m_cap_rankscore','eigen_phred','eigen_raw','epilogos_tx','is_lof','eigen_pc_raw_rankscore','epilogos_reprpcwk','fathmm_mkl_coding_rankscore','metalr_score','fathmm_mkl_coding_score','metalr_rankscore','impact_severity','metasvm_rankscore','metasvm_score','epilogos_enh','genocanyon_score','fathmm_converted_rankscore','mpc','epilogos_enhg','af_exac_all','epilogos_reprpc','max_aaf_all','mutationassessor_score','gerp','polyphen_score','gerp_elements','mutationassessor_score_rankscore','stam_mean','an_exac_all','af_exac_nfe','provean_converted_rankscore','an_exac_nfe','lrt_score','lrt_omega','grantham','lrt_converted_rankscore','genocanyon_score_rankscore','an_exac_afr','an_exac_amr','an_exac_sas','epilogos_het','ac_exac_all','linsight','gno_an_popmax','exac_num_het','an_exac_eas','gno_an_all','ac_exac_nfe','mutationtaster_converted_rankscore','an_exac_oth','an_exac_fin','gno_an_nfe','gno_af_all','gno_an_afr','epilogos_tssaflnk','gno_af_popmax','epilogos_znf','segway_sum_score','aaf_esp_ea','epilogos_txflnk','provean_score','segway_mean_score','epilogos_tss','aaf_esp_all','af_exac_amr','gno_af_nfe','epilogos_enhbiv','af_exac_sas','sift_score','fathmm_score','ac_exac_amr','aaf_esp_aa','gno_ac_all','gno_af_afr','ac_exac_sas','af_exac_eas','gno_an_fin','af_exac_afr','gno_an_eas','gno_an_oth','gno_ac_nfe','gno_ac_popmax','ac_exac_eas','ac_exac_afr','epilogos_tssbiv','gno_ac_afr','vest3_score','sigmaaf_lof_0001', 'sigmaaf_lof_01', 'sigmaaf_missense_0001', 'sigmaaf_missense_01', 'cadd_phred', 'Distill', 'primatedl'))
                               
all_processed <- gemini_ogvfb_candidates %>% 
  separate(gene_eyediseaseclass, c('RDGene','DiseaseClass'), sep='_') %>%  #split off RD disease type
  select(-RDGene) %>% 
  mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt),
         impact_severity = case_when(impact_severity == 'HIGH' ~ 3, # convert to integer 
                                     impact_severity == 'MED' ~ 2, 
                                     TRUE ~ 1),
         genesplicer = case_when(genesplicer == "" ~ 'No',
                                 grepl('^gain', genesplicer) ~ 'Gain',
                                 grepl('^loss', genesplicer) ~ 'Loss',
                                 grepl('^diff', genesplicer) ~ 'Diff',
                                 TRUE ~ 'Else')) %>% 
  mutate_at(vars(matches('ac_|an_|^n_')), funs(as.integer(.))) %>% # convert columns with ac_|whatever to integer (ac is allele count)
  mutate_at(vars(matches('af_|dann|revel|mpc|gerp|polyphen_score|sift_score|fitcons_float|gerp_elements|^adj|_z$|^pli$|^pnull$|precessive|^phylop|linsight|_rankscore$|ccr_pct_v1|linsight')), funs(as.numeric(.))) %>%  # af is allele frequency
  mutate_at(vars(one_of(numeric_predictors)), funs(as.numeric(.))) %>% 
  filter(max_aaf_all < 0.01) %>% 
  unique() # remove any common variants

# fill missing with -1
all_processed[is.na(all_processed)] <- -1


ogvfb_ML_set <- all_processed 

```
Save
```{r}
output_file <- '/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/ogvfb_exome_cohort_2018_08_07.Rdata'

save(ogvfb_ML_set, file = output_file)
```