---
title: "Build and process UK10K gemini genotypes"
output: html_notebook
---

# Load in metadata
Carss et al. supplementary
```{r}
library(tidyverse)
metadata <- readxl::read_excel(path='~/git/EGA_EGAD00001002656_NGS_reanalyze/data/1-s2.0-S0002929716305274-mmc3.xlsx') %>% mutate(Sample=Patient)
metadata
```

# Samples in gemini database
Queried with `gemini query -q 'select * from samples' EGAD00001002656.GATK.PED_EGAD00001002656.gemini.db | cut -f2`
```{r}
present_samples <- scan('~/git/EGA_EGAD00001002656_NGS_reanalyze/data/present_samples.txt', what = 'character')
```

# Create gemini query 
Didn't create gemini database with PED labelling which samples are 'solved', so I can't query on that. I'll have to build a fairly long gt-filter query including each solved sample
```{r}
library(here)
solved_samples <- metadata %>% 
  filter(Status!='Unsolved') %>% 
  filter(Sample %in% present_samples) %>% 
  pull(Sample) %>% 
  unique()
```

Let's withold 100 samples for model validation
```{r}
set.seed(8694)
withheld_samples <- solved_samples %>% sample(100)

train_samples <- setdiff(solved_samples, withheld_samples)
```

```{r}
#hom
train_gt_filter_hom <- train_samples %>% 
  paste0('gt_types.',.,' == HOM_ALT') %>% 
  paste0(., collapse = ' OR ')

train_hom_query <- paste0("gemini query --show-samples --header -q \"select * from variants where aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%'\" --gt-filter \"",
                    train_gt_filter_hom,
                    "\" EGAD00001002656.GATK.eye_var_path.PED_EGAD00001002656.gemini.db | bgzip -c > UK10K_EGAD.homs.train.gz")

#het
train_gt_filter_het <- train_samples %>% 
  paste0('gt_types.',.,' == HET') %>% 
  paste0(., collapse = ' OR ')

train_het_query <- paste0("gemini query --show-samples --header -q \"select * from variants where aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%'\" --gt-filter \"",
                    train_gt_filter_het,
                    "\" EGAD00001002656.GATK.eye_var_path.PED_EGAD00001002656.gemini.db | bgzip -c > UK10K_EGAD.hets.train.gz")

#cat(het_query)


#hom
test_gt_filter_hom <- withheld_samples %>% 
  paste0('gt_types.',.,' == HOM_ALT') %>% 
  paste0(., collapse = ' OR ')

test_hom_query <- paste0("gemini query --show-samples --header -q \"select * from variants where aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%'\" --gt-filter \"",
                    test_gt_filter_hom,
                    "\" EGAD00001002656.GATK.eye_var_path.PED_EGAD00001002656.gemini.db | bgzip -c > UK10K_EGAD.homs.test.gz")

#het
test_gt_filter_het <- withheld_samples %>% 
  paste0('gt_types.',.,' == HET') %>% 
  paste0(., collapse = ' OR ')

test_het_query <- paste0("gemini query --show-samples --header -q \"select * from variants where aaf < 0.05 and filter is NULL AND (aaf_esp_all < 0.01 AND aaf_1kg_all < 0.01 AND af_exac_all < 0.01  AND (is_coding=1 OR is_splicing=1)) OR impact_severity='HIGH' OR clinvar_sig LIKE '%patho%'\" --gt-filter \"",
                    test_gt_filter_het,
                    "\" EGAD00001002656.GATK.eye_var_path.PED_EGAD00001002656.gemini.db | bgzip -c > UK10K_EGAD.hets.test.gz")

#cat(het_query)



write('#!/bin/bash', file = here('scripts/gemini_query_calls_UK10K.sh'))
write('module load gemini', file = here('scripts/gemini_query_calls_UK10K.sh'), append = T)
write(train_het_query, file = here('scripts/gemini_query_calls_UK10K.sh'), append = T)
write(train_hom_query, file = here('scripts/gemini_query_calls_UK10K.sh'), append = T)
write(test_het_query, file = here('scripts/gemini_query_calls_UK10K.sh'), append = T)
write(test_hom_query, file = here('scripts/gemini_query_calls_UK10K.sh'), append = T)
```

# Load in result from gemini query above
```{r}
library(data.table)
test_het <- fread('gzcat /Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/UK10K/UK10K_EGAD.hets.test.gz')
test_hom <- fread('gzcat /Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/UK10K/UK10K_EGAD.homs.test.gz')

train_het <- fread('gzcat /Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/UK10K/UK10K_EGAD.hets.train.gz')
train_hom <- fread('gzcat /Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/data/UK10K/UK10K_EGAD.homs.train.gz')
```

# Create regions file
The regions file is used to generate coordinates for the selected variants, which can be used by `tabix --regions` to create a cut-down UK10K vcf to more quickly iterate gemini database / annotation tweaks. The full vcf takes ~5 days to re-annotate

I've gone up into `# Create gemini query` and changed the database to the cut down one I made
```{r}
regions_f <- rbind(test_het %>% mutate(region=paste0(chrom,'\t',end,'\t',end)),
                   test_hom %>% mutate(region=paste0(chrom,'\t',end,'\t',end)),
                   train_het %>% mutate(region=paste0(chrom,'\t',end,'\t',end)),
                   train_hom %>% mutate(region=paste0(chrom,'\t',end,'\t',end))) %>% 
  mutate(end=as.integer(end)) %>% 
  arrange(chrom, end) %>% 
  select(region)

write(regions_f$region, file='/Volumes/data/projects/nei/mcgaughey/EGA_EGAD00001002656_7n/vcf_annotated_v2/eye_var_regions.txt')
# run on biowulf2
# tabix EGAD00001002656.GATK.vcf.gz --print-header --regions eye_var_regions.txt | bgzip -c > /data/mcgaugheyd/projects/nei/mcgaughey/eye_var_Pathogenicity/data/EGAD00001002656.GATK.eye_var_path.vcf.gz &
# tabix -p vcf /data/mcgaugheyd/projects/nei/mcgaughey/eye_var_Pathogenicity/data/EGAD00001002656.GATK.eye_var_path.vcf.gz
```

# Next step
run scripts/build_10K_data.R
