---
title: "Model Assesement"
output: 
  html_notebook:
    theme: flatly
    toc: True
---

# Load Data
```{r}
library(tidyverse)
library(caret)
library(ModelMetrics)
library(randomForest)
library(keras)
library(tensorflow)
library(PRROC)
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/VPaC_12mtry_v11.Rdata')
load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/allX_2018_08_03.Rdata')
allX$primatedl  <- as.numeric(allX$primatedl)
allX[is.na(allX)] <- -1

load('/Volumes/data/projects/nei/mcgaughey/eye_var_Pathogenicity/clean_data/assess_2018_08_03.Rdata')
assess_set$primatedl  <- as.numeric(assess_set$primatedl)
assess_set[is.na(assess_set)] <- -1
#allX <- allX %>% mutate_at(vars(contains('VP')), as.numeric) %>% mutate_at(vars(contains('Deep')), as.numeric)
# confusion matrix maker for built models
cm_maker <- function(predictor = 'cadd_phred', data, cutoff=0.5, mode = 'prec_recall') {
  if (class(predictor)!='character'){
    print("Running in predictor is a model mode")
    new_predictions <- predict(predictor, data, type='prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Answers == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = as.factor(new_predictions$Prediction), reference = as.factor(new_predictions$Answers), mode= mode)
    out$MCC <- mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff)
  } else {
    print("Running in predictor is a precomputed column in data mode")
    new_predictions <- data 
    new_predictions$Prediction <- 'NotPathogenic'
    new_predictions[(new_predictions[,predictor] > cutoff), 'Prediction'] <- "Pathogenic"
    new_predictions <- new_predictions %>% mutate(preds = case_when(Prediction == 'Pathogenic' ~ 1,
                                                                    TRUE ~ 0),
                                                  actuals = case_when(Status == 'Pathogenic' ~ 1,
                                                                      TRUE ~ 0))
    out <- caret::confusionMatrix(data = as.factor(new_predictions$Prediction), reference = as.factor(new_predictions$Status), mode= mode)
    out$MCC <- mcc(new_predictions$preds, new_predictions$actuals, cutoff=cutoff)
  }
  out
}


```

# Table Data
```{r}

```

# Correlation of predictors
## All predictors
```{r, fig.width=10, fig.height=10}
library(superheat)

numeric_predictors<-c('is_exonic','is_coding','is_lof','is_splicing','impact_severity','polyphen_score','sift_score','dann','eigen_phred','fathmm_converted_rankscore','gerp','genocanyon_score_rankscore','hgmd_overlap','linsight','lrt_omega','lrt_converted_rankscore','m_cap_rankscore','mpc','metalr_rankscore','metasvm_rankscore','mutationassessor_score_rankscore','mutationtaster_converted_rankscore','provean_converted_rankscore','provean_score','revel','vest3_rankscore','aaf_1kg_afr','aaf_1kg_all','aaf_1kg_amr','aaf_1kg_eas','aaf_1kg_eur','aaf_1kg_sas','aaf_esp_aa','aaf_esp_all','aaf_esp_ea','ac_exac_afr','ac_exac_all','ac_exac_amr','ac_exac_eas','ac_exac_fin','ac_exac_nfe','ac_exac_oth','ac_exac_sas','adj_exp_lof','adj_exp_mis','adj_exp_syn','af_exac_afr','af_exac_all','af_exac_amr','af_exac_eas','af_exac_nfe','af_exac_oth','af_exac_sas','an_exac_afr','an_exac_all','an_exac_amr','an_exac_eas','an_exac_fin','an_exac_nfe','an_exac_oth','an_exac_sas','ccr_pct_v1','cpg_island','epilogos_bivflnk','epilogos_enh','epilogos_enhbiv','epilogos_enhg','epilogos_het','epilogos_quies','epilogos_reprpc','epilogos_reprpcwk','epilogos_tss','epilogos_tssaflnk','epilogos_tssbiv','epilogos_tx','epilogos_txflnk','epilogos_txwk','epilogos_znf','exac_num_het','exac_num_hom_alt','fathmm_mkl_coding_rankscore','fitcons','geno2mp','gerp_elements','gno_ac_afr','gno_ac_all','gno_ac_amr','gno_ac_asj','gno_ac_eas','gno_ac_fin','gno_ac_nfe','gno_ac_oth','gno_ac_popmax','gno_af_afr','gno_af_all','gno_af_amr','gno_af_asj','gno_af_eas','gno_af_fin','gno_af_nfe','gno_af_oth','gno_af_popmax','gno_an_afr','gno_an_all','gno_an_amr','gno_an_asj','gno_an_eas','gno_an_fin','gno_an_nfe','gno_an_oth','gno_an_popmax','gno_id','gno_popmax','hapmap1','hapmap2','in_1kg','in_esp','in_exac','lof_z','max_aaf_all','mis_z','n_lof','n_mis','n_syn','pli','pnull','precessive','phylop_100way','segway_mean_score','segway_sum_score','stam_mean','syn_z','grantham','cadd_phred', 'sigmaaf_lof_0001', 'sigmaaf_lof_01', 'sigmaaf_missense_0001', 'sigmaaf_missense_01')


heatmap_predictors<-c('impact_severity','polyphen_score','sift_score','dann','eigen_phred','fathmm_converted_rankscore','gerp','genocanyon_score_rankscore','linsight','lrt_omega','lrt_converted_rankscore','m_cap_rankscore','mpc','metalr_rankscore','metasvm_rankscore','mutationassessor_score_rankscore','mutationtaster_converted_rankscore','provean_converted_rankscore','provean_score','revel','vest3_rankscore','aaf_1kg_afr','aaf_1kg_all','aaf_1kg_amr','aaf_1kg_eas','aaf_1kg_eur','aaf_1kg_sas','aaf_esp_aa','aaf_esp_all','aaf_esp_ea','ac_exac_afr','ac_exac_all','ac_exac_amr','ac_exac_eas','ac_exac_fin','ac_exac_nfe','ac_exac_oth','ac_exac_sas','adj_exp_lof','adj_exp_mis','adj_exp_syn','af_exac_afr','af_exac_all','af_exac_amr','af_exac_eas','af_exac_nfe','af_exac_oth','af_exac_sas','an_exac_afr','an_exac_all','an_exac_amr','an_exac_eas','an_exac_fin','an_exac_nfe','an_exac_oth','an_exac_sas','ccr_pct_v1','cpg_island','epilogos_bivflnk','epilogos_enh','epilogos_enhbiv','epilogos_enhg','epilogos_het','epilogos_quies','epilogos_reprpc','epilogos_reprpcwk','epilogos_tss','epilogos_tssaflnk','epilogos_tssbiv','epilogos_tx','epilogos_txflnk','epilogos_txwk','epilogos_znf','exac_num_het','exac_num_hom_alt','fathmm_mkl_coding_rankscore','fitcons','geno2mp','gerp_elements','gno_ac_afr','gno_ac_all','gno_ac_amr','gno_ac_asj','gno_ac_eas','gno_ac_fin','gno_ac_nfe','gno_ac_oth','gno_ac_popmax','gno_af_afr','gno_af_all','gno_af_amr','gno_af_asj','gno_af_eas','gno_af_fin','gno_af_nfe','gno_af_oth','gno_af_popmax','gno_an_afr','gno_an_all','gno_an_amr','gno_an_asj','gno_an_eas','gno_an_fin','gno_an_nfe','gno_an_oth','gno_an_popmax','gno_id','gno_popmax','hapmap1','hapmap2','in_1kg','in_esp','in_exac','lof_z','max_aaf_all','mis_z','n_lof','n_mis','n_syn','pli','pnull','precessive','phylop_100way','segway_mean_score','segway_sum_score','stam_mean','syn_z','grantham','cadd_phred', 'sigmaaf_lof_0001', 'sigmaaf_lof_01', 'sigmaaf_missense_0001', 'sigmaaf_missense_01', 'Distill')

# remove near zero var predictors with caret's nearZeroVar
dat <- assess_set %>% filter(DataSet == 'SuperGrimm') %>% select(one_of(numeric_predictors, 'Distill','primatedl')) %>%  mutate_at(vars(one_of(numeric_predictors, 'Distill','primatedl')), funs(as.numeric(.))) %>% 
  select(heatmap_predictors)
#rem <- nearZeroVar(dat)
#dat <- dat[,-rem]
superheat(cor(dat),
          pretty.order.rows = T, 
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, 
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.2)
```

## Removing population AF
```{r, fig.width=7, fig.height=5.5}

heatmap_predictors<-c('impact_severity','polyphen_score','sift_score','dann','eigen_phred','fathmm_converted_rankscore','gerp','genocanyon_score_rankscore','linsight','lrt_omega','lrt_converted_rankscore','m_cap_rankscore','mpc','metalr_rankscore','metasvm_rankscore','mutationassessor_score_rankscore','mutationtaster_converted_rankscore','provean_converted_rankscore','provean_score','revel','vest3_rankscore','ccr_pct_v1','cpg_island','epilogos_bivflnk','epilogos_enh','epilogos_enhbiv','epilogos_enhg','epilogos_het','epilogos_quies','epilogos_reprpc','epilogos_reprpcwk','epilogos_tss','epilogos_tssaflnk','epilogos_tssbiv','epilogos_tx','epilogos_txflnk','epilogos_txwk','epilogos_znf','exac_num_het','exac_num_hom_alt','fathmm_mkl_coding_rankscore','fitcons','geno2mp','gerp_elements', 'hapmap1','hapmap2','in_1kg','in_esp','in_exac','lof_z','max_aaf_all','mis_z','n_lof','n_mis','n_syn','pli','pnull','precessive','phylop_100way','segway_mean_score','segway_sum_score','stam_mean','syn_z','grantham','cadd_phred', 'sigmaaf_lof_0001', 'sigmaaf_lof_01', 'sigmaaf_missense_0001', 'sigmaaf_missense_01', 'Distill','primatedl')

# remove near zero var predictors with caret's nearZeroVar
dat <- assess_set %>% filter(DataSet == 'SuperGrimm') %>% select(one_of(numeric_predictors, 'Distill','primatedl')) %>%  mutate_at(vars(one_of(numeric_predictors, 'Distill','primatedl')), funs(as.numeric(.))) %>% 
  select(heatmap_predictors) 
colnames(dat) <- gsub('_rankscore|_score|_converted','',colnames(dat))
#rem <- nearZeroVar(dat)
#dat <- dat[,-rem]
superheat(cor(dat),
          pretty.order.rows = T, row.dendrogram = T,
          pretty.order.cols = T, 
          left.label = 'variable', 
          force.left.label=T, 
          bottom.label = 'none', 
          left.label.text.alignment = 'right', 
          left.label.text.size = 4, 
          left.label.size = 0.2)
```

# Variable Importance
For random forest model

Mean Decrease Gini
```{r, fig.height=1.5, fig.width=7}
VPaC_12mtry_v11$importance %>% data.frame() %>% rownames_to_column('Predictor') %>% 
  arrange(MeanDecreaseGini) %>% 
  tail(20) %>% 
  mutate(Predictor = factor(Predictor, levels=unique(Predictor))) %>% 
  ggplot(aes(x=Predictor, y=MeanDecreaseGini)) +
  geom_point() +
  coord_flip() +
  theme_minimal()
```

Mean Decrease Accuracy
```{r, fig.height=1.5, fig.width=7}
VPaC_12mtry_v11$importance %>% data.frame() %>% rownames_to_column('Predictor') %>% 
  arrange(MeanDecreaseAccuracy) %>% 
  tail(20) %>% 
  mutate(Predictor = factor(Predictor, levels=unique(Predictor))) %>% 
  ggplot(aes(x=Predictor, y=MeanDecreaseAccuracy)) +
  geom_point() +
  coord_flip() +
  theme_minimal()
```


# AUC-PR AUC-ROC Calculation Functions
```{r, fig.width=5, fig.height=1.5}
library(PRROC)
library(cowplot)

# precision recall AUC
pr_maker <- function(predictor, data, cutoff=0.5) {
  
  if (class(predictor)=='character'){
    predictor <- enquo(predictor)
    pr.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
             scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
             curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    pr.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
             scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
             curve = T)
  }
}

# flip path/not path
pr_makerX <- function(predictor, data, cutoff=0.5) {
  
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    pr.curve(scores.class0 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
             scores.class1 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
             curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    pr.curve(scores.class0 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
             scores.class1 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
             curve = T)
  }
}

# ROC AUC
roc_maker <- function(predictor, data, cutoff=0.5) {
  
  if (class(predictor)=='character'){
    predictor = enquo(predictor)
    roc.curve(scores.class0 = data %>% filter(Status=='Pathogenic') %>% pull(!!predictor),
              scores.class1 = data %>% filter(Status=='NotPathogenic') %>% pull(!!predictor),
              curve = T)
  }
  else {
    new_predictions <- predict(predictor, data, type = 'prob') %>% data.frame() %>% 
      mutate(Answers = data$Status, Prediction = case_when(Pathogenic > cutoff ~ 'Pathogenic', TRUE ~ 'NotPathogenic'))
    roc.curve(scores.class0 = new_predictions %>% filter(Answers=='Pathogenic') %>% pull(Pathogenic),
              scores.class1 = new_predictions %>% filter(Answers=='NotPathogenic') %>% pull(Pathogenic),
              curve = T)
  }
}
```


# Assess Performance of Models and Existing Predictors in Assess Sets
```{r}

model_vec <- c('DeepRNN','VPaC', 'Distill', 'xgbTree','revel','cadd_phred', 'primatedl', 'fitcons', 'metalr_rankscore', 'fathmm_converted_rankscore', 'm_cap_rankscore')
pr_data <- data.frame()
roc_data <- data.frame()

for (i in model_vec){
  #print(i)
  for (j in unique(assess_set %>% filter(DataSet != 'Extra NotPathogenic') %>% pull(DataSet))){
    #print(j)
    # Samocha has way more path than not path, so use flipped pr_maker
    if (j=='Samocha'){
      pr <- pr_makerX(i, assess_set %>% filter(DataSet == j))
    } else{
      pr <- pr_maker(i, assess_set %>% filter(DataSet == j))
    }
    out <- pr$curve[,1:2] %>% data.frame()
    colnames(out) <- c('Recall','Precision')
    #cm_out <- cm_maker(i, assess_set %>% filter(DataSet == j))
    out$AUC <- pr$auc.integral
    out$Model <- i
    #out$MCC <- cm_out$MCC
    out$'Model (AUC)' <- paste0(i, ' (',round(pr$auc.integral,5),')' )
    out$DataSet <- j
    pr_data <- rbind(pr_data, out)
  }
}

for (i in model_vec){
  #print(i)
  for (j in unique(assess_set$DataSet)){
    roc <- roc_maker(i, assess_set %>% filter(DataSet == j))
    out <- roc$curve[,1:2] %>% data.frame()
    colnames(out) <- c('FPR','Sensitivity')
    out$AUC <- roc$auc
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc,5),')' )
    out$DataSet <- j
    roc_data <- rbind(roc_data, out)
  }
}
#rm(assess_set)
```
## Table Data
```{r}
pr_table <- pr_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% 
  dplyr::select(AUC, Model, DataSet) %>% 
  filter(grepl('Grimm|Test|UniFun|DDL|Sam|Hom|With', DataSet)) %>% 
  unique() %>% 
  mutate(AUC=round(AUC,4)) %>% 
  spread(Model, AUC) %>% 
  DT::datatable()

roc_table <- roc_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% 
  dplyr::select(AUC, Model, DataSet) %>% 
  filter(grepl('Grimm|Test|UniFun|DDL|Sam|Hom|With', DataSet)) %>% 
  unique() %>% 
  mutate(AUC=round(AUC,4)) %>% 
  spread(Model, AUC) %>% 
  DT::datatable()

pr_table

roc_table
```
## Dot plots of AUC
```{r, fig.height=6}
# pr_data %>% dplyr::select(AUC, Model, DataSet) %>%
#   #filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>%
#   unique() %>%
#   ggplot(aes(x=Model, y=AUC)) +
#   geom_point(stat = 'identity') +
#   coord_cartesian(ylim=c(0,1)) +
#   facet_wrap(~DataSet, nrow=2) +
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('Precision Recall')
# 
# roc_data %>% dplyr::select(AUC, Model, DataSet) %>%
#   #filter(grepl('Grimm|ClinVar|Test|UniFun|DDL|Sam|Hom', DataSet)) %>%
#   unique()%>%
#   ggplot(aes(x=Model, y=AUC)) +
#   geom_point(stat = 'identity') + coord_cartesian(ylim=c(0,1)) + facet_wrap(~DataSet, nrow=2) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle('Receiver Operating Characteristic')
```

## Homsy and Samocha ROC
```{r, fig.height=3, fig.width=3}
sam <- roc_data %>%  
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('Samoch', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=`Model (AUC)`)) + 
  geom_line() + 
  theme_minimal() + 
  ggtitle('Samocha') +
  ggsci::scale_color_futurama()

homs <- roc_data %>%  
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('Homs', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=`Model (AUC)`)) + 
  geom_line() + 
  theme_minimal() + 
  ggtitle('Homsy') +
  ggsci::scale_color_futurama()

title <- ggdraw() + draw_label("Area Under the ROC Curve")
cowplot::plot_grid(homs, sam, ncol = 1)
```

## SuperGrimm, UK10K, DDL NISC RD Cohort, 
```{r, fig.height=6, fig.width=3}
uk10k <- pr_data %>% 
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('With', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + 
  geom_step() + 
  theme_minimal() + 
  ggsci::scale_color_futurama() + 
  ggtitle('Carss IRD') 

ddl <- pr_data %>% 
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('DDL', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + 
  geom_step() + 
  theme_minimal() + 
  ggsci::scale_color_futurama() + 
  ggtitle('NEI IRD') 

superG <- pr_data %>% 
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('Super', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + 
  geom_step() + 
  theme_minimal() + 
  ggsci::scale_color_futurama() + 
  ggtitle('Super Grimm') 

test_set <- pr_data %>% 
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  filter(grepl('Test Set', DataSet)) %>% 
  filter(!grepl('Deep|VPaC|xgb', Model)) %>% 
  ggplot(aes(x=Recall, y=Precision, colour=`Model (AUC)`)) + 
  geom_step() + 
  theme_minimal() + 
  ggsci::scale_color_futurama() + 
  ggtitle('Test Data') 
  
title <- ggdraw() + draw_label("Area Under the Precision Recall Curves")
cowplot::plot_grid(title, uk10k, superG, ddl, test_set, ncol=1)


```




# Assess Performance of Models and Existing Predictors across Grimm sets

## Counts
```{r}
allX %>% filter(grepl('Grimm', DataSet)) %>% group_by(DataSet, Status) %>% summarise(Count = n())
```
## Calculate
```{r}
model_vec <- c('DeepRNN','VPaC', 'Distill', 'xgbTree','revel','cadd_phred', 'primatedl', 'fitcons', 'metalr_rankscore', 'fathmm_converted_rankscore', 'm_cap_rankscore')
Gpr_data <- data.frame()
Groc_data <- data.frame()

for (i in model_vec){
  print(i)
  for (j in unique(allX %>% filter(grepl('Grimm', DataSet)) %>% pull(DataSet))){
    print(j)
   
    pr <- pr_maker(i, allX %>% filter(DataSet == j))

    out <- pr$curve[,1:2] %>% data.frame()
    colnames(out) <- c('Recall','Precision')
    #cm_out <- cm_maker(i, allX %>% filter(DataSet == j))
    out$AUC <- pr$auc.integral
    out$Model <- i
    #out$MCC <- cm_out$MCC
    out$'Model (AUC)' <- paste0(i, ' (',round(pr$auc.integral,5),')' )
    out$DataSet <- j
    Gpr_data <- rbind(Gpr_data, out)
  }
}

for (i in model_vec){
  #print(i)
  for (j in unique(allX %>% filter(grepl('Grimm', DataSet)) %>% pull(DataSet))){
    roc <- roc_maker(i, allX %>% filter(DataSet == j))
    out <- roc$curve[,1:2] %>% data.frame()
    colnames(out) <- c('FPR','Sensitivity')
    out$AUC <- roc$auc
    out$Model <- i
    out$'Model (AUC)' <- paste0(i, ' (',round(roc$auc,5),')' )
    out$DataSet <- j
    Groc_data <- rbind(Groc_data, out)
  }
}
#rm(allX)
```

## Table Data
```{r}
Gpr_table <- Gpr_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% 
  dplyr::select(AUC, Model, DataSet) %>% 
  #filter(grepl('Grimm|Test|UniFun|DDL|Sam|Hom|With', DataSet)) %>% 
  unique() %>% 
  mutate(AUC=round(AUC,4)) %>% 
  spread(Model, AUC) %>% 
  DT::datatable()

Groc_table <- Groc_data %>% mutate(Model = factor(Model, levels = model_vec)) %>% 
  dplyr::select(AUC, Model, DataSet) %>% 
  #filter(grepl('Grimm|Test|UniFun|DDL|Sam|Hom|With', DataSet)) %>% 
  unique() %>% 
  mutate(AUC=round(AUC,4)) %>% 
  spread(Model, AUC) %>% 
  DT::datatable()

Gpr_table

Groc_table
```

## AUC 
Unimpressive performance for Distill
```{r}
Groc_data %>%
  mutate(`Model (AUC)` = (paste0(Model, ' (', round(AUC, 2),')'))) %>% 
  #filter(DataSet == 'ExoVar') %>% 
  filter(!grepl('Deep|VPaC', Model)) %>% 
  ggplot(aes(x=FPR, y=Sensitivity, colour=Model)) +
  facet_wrap(~DataSet) +
  geom_line() + 
  theme_minimal() + 
  ggsci::scale_color_futurama()
```


# Find optimal cutoff for DataSet X for model Y
Plot cutoff against MCC
```{r}
mcc_view <- function(sequence = seq(0,1,by = 0.01), predictor = 'Distill', assess_set = assess_set %>% filter(grepl('UK10', DataSet))){
  out <- data.frame(cutoff = sequence)
  mccs <- ''
  for (i in sequence){
    capture.output(mccs <- c(mccs, cm_maker(predictor, assess_set, i)$MCC), file='/dev/null')
  }
  out$mcc <- mccs[2:length(mccs)]
  out$mcc[out$mcc == 'NaN'] <- 0
  out$mcc <- as.numeric(out$mcc)
  out$cutoff <- as.numeric(out$cutoff)
  out
}
print('Distill, DDL')
suppressWarnings(mcc_view(predictor='Distill', assess_set = assess_set %>% filter(grepl('DDL', DataSet))) %>% arrange(-mcc) %>% head())
print('Distill, UK10K')
suppressWarnings(mcc_view(predictor='Distill', assess_set = assess_set %>% filter(grepl('UK10K', DataSet))) %>% arrange(-mcc) %>% head())
```
# OGVFB Exomes
```{r}
ogvfb$ogvfb_ML_set$VPaC <-predict(VPaC_6mtry, ogvfb$ogvfb_ML_set, type='prob')[,1]
ogvfb$ogvfb_ML_set$DiseaseClass <- factor(ogvfb$ogvfb_ML_set$DiseaseClass, levels=c('-1','Albinism','Albinism,Albinism','Developmental','RD','RD,Developmental','Stargardt,RD'))
ogvfb$ogvfb_ML_set$OVPaC <-predict(OVPaC_6mtry, ogvfb$ogvfb_ML_set, type='prob')[,1]

# OVPaC most path
ogvfb$ogvfb_ML_set %>% filter(OVPaC > 0.1) %>% 
  left_join(., ogvfb$gemini_ogvfb_candidates %>%  
              mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt), hgvsc=gsub('.*:','',hgvsc), hgvsp=gsub('.*:','',hgvsp)) %>% 
              select(pos_id, contains('hgvs'), contains('family'), gene, impact_so, num_het, num_hom_alt, filter, hgmd_overlap, clinvar_pathogenic, pubmed)) %>% 
  filter(num_het < 4, num_hom_alt < 4, filter == 'None') %>% 
  select(pos_id, gene, hgvsc, hgvsp, impact_so, hgmd_overlap, clinvar_pathogenic, pubmed, VPaC, OVPaC, ccr_pct_v1, gno_an_popmax, cadd_phred, fitcons_float, revel, family_id, family_members, family_genotypes) %>% 
  mutate(family_genotypes = gsub(',','<br/>', family_genotypes), family_members = gsub(',','<br/>', family_members)) %>% 
  arrange(-OVPaC) %>% 
  mutate_if(is.numeric, funs(round(.,2))) %>% 
  DT::datatable(escape = FALSE, rownames = FALSE)

# VPaC most path
ogvfb$ogvfb_ML_set %>% filter(VPaC > 0.1) %>% 
  left_join(., ogvfb$gemini_ogvfb_candidates %>%  
              mutate(pos_id=paste0(chrom, ':', end, '_', ref, '_', alt), hgvsc=gsub('.*:','',hgvsc), hgvsp=gsub('.*:','',hgvsp)) %>% 
              select(pos_id, contains('hgvs'), contains('family'), gene, impact_so, num_het, num_hom_alt, filter, hgmd_overlap, clinvar_pathogenic, pubmed)) %>% 
  filter(num_het < 4, num_hom_alt < 4, filter == 'None') %>% 
  select(pos_id, gene, hgvsc, hgvsp, impact_so, hgmd_overlap, clinvar_pathogenic, pubmed, VPaC, OVPaC, ccr_pct_v1, gno_an_popmax, cadd_phred, fitcons_float, revel, family_id, family_members, family_genotypes) %>% 
  mutate(family_genotypes = gsub(',','<br/>', family_genotypes), family_members = gsub(',','<br/>', family_members)) %>% 
  arrange(-VPaC) %>% 
  mutate_if(is.numeric, funs(round(.,2))) %>% 
  DT::datatable(escape = FALSE, rownames = FALSE)
```

